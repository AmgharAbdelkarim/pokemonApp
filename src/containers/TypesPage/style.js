import styled from "styled-components";

export const TypeWrapper = styled.div`
  width: 70%;
  display: grid;
  margin: 0 auto;
  .item_wrapper {
    display: flex;
    justify-content: space-evenly;
    flex-wrap: wrap;
    margin-bottom: 30px;
  }
  .button_wrapper {
    display: flex;
    justify-content: center;
    margin: 20px 0;
  }
`;
