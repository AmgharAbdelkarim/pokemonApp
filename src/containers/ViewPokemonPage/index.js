import React, { useEffect } from "react";

import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Card from "components/Card";
import { PokemonContainer } from "containers/ViewPokemonPage/styles";
import Box from "components/Box";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonDetailsSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getPokemonEvolutionSelector,
} from "store/selectors";
import { getPokemonDetails, getPokemonEvolutions } from "store/action";

const ViewPokemon = ({
  match,
  history,
  pokemon,
  getPokemonDetails,
  getPokemonEvolutions,
  evolution,
  loading,
  hasError,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonDetails(params.name);
  }, [params.name]);
  useEffect(() => {
    getPokemonEvolutions(pokemon.id);
  }, [pokemon]);
  return (
    <PokemonContainer id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.name ? (
        <React.Fragment>
          <Card pokemon={pokemon} clickHandler={(url) => history.push(url)} />
          {evolution.length > 0 && (
            <React.Fragment>
              <h3>pokemon evolutions</h3>
              {evolution.map(({ species }, index) => (
                <Box
                  key={index}
                  name={species.name}
                  clickHandler={() => history.push(`/pokemon/${species.name}`)}
                />
              ))}
            </React.Fragment>
          )}
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </PokemonContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonDetailsSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  evolution: getPokemonEvolutionSelector,
});

const mapDispatchToProps = {
  getPokemonDetails,
  getPokemonEvolutions,
};

export default connect(mapStateToProps, mapDispatchToProps)(ViewPokemon);
export { ViewPokemon };
