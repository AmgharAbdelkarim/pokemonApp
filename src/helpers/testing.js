import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import store from "store";
import { Router as Switch } from "react-router-dom";

import { createBrowserHistory } from "history";

export const history = createBrowserHistory();

export const Wrapper = (props) => {
  return (
    <Provider store={store()}>
      <Switch history={history}>{props.children}</Switch>
    </Provider>
  );
};

export const renderWithRedux = (ui, renderOptions = {}) => {
  return {
    ...render(ui, {
      wrapper: Wrapper,
      ...renderOptions,
    }),
    store,
  };
};
