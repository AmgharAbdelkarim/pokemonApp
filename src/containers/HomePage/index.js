import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { ContentWrapper } from "containers/HomePage/style";
import Pokemon from "components/Box";
import Button from "components/Button";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonSelector,
  getSizeSelector,
  getTypesSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getCurrentUrlSelector,
  getNextUrlSelector,
  getPrevUrlSelector,
  getSelectSelectedTypes,
} from "store/selectors";
import {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
} from "store/action";

const Home = ({
  pokemon,
  size,
  types,
  loading,
  hasError,
  currentUrl,
  nextUrl,
  prevUrl,
  getPokemon,
  selectedTypes,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
  history,
}) => {
  useEffect(() => {
    getPokemon();
  }, [currentUrl]);

  useEffect(() => {
    getTypes();
  }, []);

  const nextHandler = () => {
    if (!nextUrl) return setSize(size + 20);
    setCurrentUrl(nextUrl);
  };

  const prevHandler = () => {
    if (!nextUrl) return setSize(size - 20);
    setCurrentUrl(prevUrl);
  };

  const filterByType = (currentUrl, name) => {
    setCurrentUrl(currentUrl);
    setTypeSelected(name);
  };

  return (
    <ContentWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="filter_box">
            <h5>filter Pokemon By Type</h5>
            {selectedTypes.length > 0 && (
              <Button
                className="clear_button"
                clickHandler={() => {
                  setCurrentUrl("https://pokeapi.co/api/v2/pokemon");
                  clearTypeSelected();
                }}
              >
                Clear
              </Button>
            )}
            <ul>
              {types.map(({ name, url }) => (
                <li
                  onClick={() => filterByType(url, name)}
                  aria-hidden="true"
                  key={name}
                  className={selectedTypes.indexOf(name) >= 0 ? "selected" : ""}
                >
                  <span>{name}</span>
                </li>
              ))}
            </ul>
          </div>
          <div className="item_container">
            <div className="item_wrapper">
              {pokemon.slice(size, size + 20).map((p, index) => (
                <Pokemon
                  key={index}
                  name={p.name || p.pokemon.name}
                  clickHandler={() =>
                    history.push(`/pokemon/${p.name || p.pokemon.name}`)
                  }
                />
              ))}
            </div>
            <div className="button_wrapper">
              <Button
                disabled={prevUrl || size > 19 ? false : true}
                clickHandler={() => prevHandler()}
              >
                previous
              </Button>
              <Button
                disabled={nextUrl || size < pokemon.length - 19 ? false : true}
                clickHandler={() => nextHandler()}
              >
                next
              </Button>
            </div>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </ContentWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonSelector,
  size: getSizeSelector,
  types: getTypesSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  currentUrl: getCurrentUrlSelector,
  nextUrl: getNextUrlSelector,
  prevUrl: getPrevUrlSelector,
  selectedTypes: getSelectSelectedTypes,
});

const mapDispatchToProps = {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
};
export default connect(mapStateToProps, mapDispatchToProps)(Home);

export { Home };
