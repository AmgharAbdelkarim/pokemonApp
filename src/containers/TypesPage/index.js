import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Pokemon from "components/Box";
import Button from "components/Button";
import { TypeWrapper } from "containers/TypesPage/style";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonTypesSelector,
  getSizeSelector,
  getLoadingSelector,
  getHasErrorSelector,
} from "store/selectors";
import { getPokemonTypes, setSize } from "store/action";

const Types = ({
  match,
  pokemon,
  size,
  loading,
  hasError,
  getPokemonTypes,
  setSize,
  history,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonTypes(params.type);
  }, []);
  const nextHandler = () => {
    setSize(size + 20);
  };
  const prevHandler = () => {
    setSize(size - 20);
  };
  return (
    <TypeWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="item_wrapper">
            {pokemon.slice(size, size + 20).map(({ pokemon }, index) => (
              <Pokemon
                clickHandler={() => history.push(`/pokemon/${pokemon.name}`)}
                name={pokemon.name}
                key={index}
              />
            ))}
          </div>
          <div className="button_wrapper">
            <Button
              disabled={size > 19 ? false : true}
              clickHandler={() => prevHandler()}
            >
              previous
            </Button>
            <Button
              disabled={size < pokemon.length - 19 ? false : true}
              clickHandler={() => nextHandler()}
            >
              next
            </Button>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </TypeWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonTypesSelector,
  size: getSizeSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
});

const mapDispatchToProps = {
  getPokemonTypes,
  setSize,
};

export default connect(mapStateToProps, mapDispatchToProps)(Types);

export { Types };
