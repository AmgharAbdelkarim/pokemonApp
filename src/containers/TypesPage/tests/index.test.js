import { cleanup, fireEvent, waitFor } from "@testing-library/react";

import { renderWithRedux } from "helpers/testing";

import { Types } from "containers/TypesPage";
import { createBrowserHistory } from "history";

afterEach(cleanup);

describe("Types", () => {
  let props;
  beforeEach(() => {
    props = {
      pokemon: [
        {
          pokemon: {
            name: "pidgey",
            url: "https://pokeapi.co/api/v2/pokemon/16/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "pidgeotto",
            url: "https://pokeapi.co/api/v2/pokemon/17/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "pidgeot",
            url: "https://pokeapi.co/api/v2/pokemon/18/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "rattata",
            url: "https://pokeapi.co/api/v2/pokemon/19/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "raticate",
            url: "https://pokeapi.co/api/v2/pokemon/20/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "spearow",
            url: "https://pokeapi.co/api/v2/pokemon/21/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "fearow",
            url: "https://pokeapi.co/api/v2/pokemon/22/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "jigglypuff",
            url: "https://pokeapi.co/api/v2/pokemon/39/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "wigglytuff",
            url: "https://pokeapi.co/api/v2/pokemon/40/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "meowth",
            url: "https://pokeapi.co/api/v2/pokemon/52/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "persian",
            url: "https://pokeapi.co/api/v2/pokemon/53/",
          },
          slot: 1,
        },
        {
          pokemon: {
            name: "farfetchd",
            url: "https://pokeapi.co/api/v2/pokemon/83/",
          },
          slot: 1,
        },
      ],
      size: 0,
      loading: false,
      hasError: false,
      getPokemonTypes: jest.fn(),
      setSize: jest.fn(),
      match: {
        params: "fakeType",
      },
    };
  });
  afterEach(cleanup);
  it("should Types render correctly", () => {
    const { asFragment } = renderWithRedux(<Types {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render loading state if loading is true", () => {
    let newProps = { ...props, loading: true };
    const { getByText } = renderWithRedux(<Types {...newProps} />);
    expect(getByText("Loading...")).toBeInTheDocument();
  });
  it("should render error state if hasError is true", () => {
    let newProps = { ...props, hasError: true };
    const { getByText } = renderWithRedux(<Types {...newProps} />);
    expect(getByText("Something Went Wrong")).toBeInTheDocument();
  });
  it("should call setSize action when we click on nextButton", async () => {
    const { getByRole } = renderWithRedux(<Types {...props} size={-20} />);
    const nextButton = getByRole("button", { name: /next/ });
    fireEvent.click(nextButton);
    await waitFor(() => expect(props.setSize).toHaveBeenCalledWith(0));
  });
  it("should call setSize action when we click on prevButton", async () => {
    const { getByRole } = renderWithRedux(<Types {...props} size={20} />);
    const prevButton = getByRole("button", { name: /previous/ });
    fireEvent.click(prevButton);
    await waitFor(() => expect(props.setSize).toHaveBeenCalledWith(0));
  });
  it("should next button be disabled", () => {
    let newProps = { ...props, size: 20 };
    const { getByRole } = renderWithRedux(<Types {...newProps} />);
    const nextButton = getByRole("button", { name: /next/ });
    expect(nextButton).toBeDisabled();
  });
  it("should next button be disabled", async () => {
    let newProps = { ...props, size: 0 };
    const { getByRole } = renderWithRedux(<Types {...newProps} />);
    const prevButton = getByRole("button", { name: /previous/ });
    expect(prevButton).toBeDisabled();
  });

  it("should redirect to view pokemon page if i click on pokemon", async () => {
    const history = createBrowserHistory();
    jest.spyOn(history, "push");
    const { getByText } = renderWithRedux(
      <Types {...props} history={history} />
    );
    const pokemon = getByText("pidgey");
    fireEvent.click(pokemon);
    await waitFor(() =>
      expect(history.push).toHaveBeenCalledWith("/pokemon/pidgey")
    );
  });
});
