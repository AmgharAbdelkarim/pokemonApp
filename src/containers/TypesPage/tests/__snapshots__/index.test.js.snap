// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types should Types render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-dlfnbm eEFYSP"
    id="content"
  >
    <div
      class="item_wrapper"
    >
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          pidgey
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          pidgeotto
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          pidgeot
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          rattata
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          raticate
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          spearow
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          fearow
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          jigglypuff
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          wigglytuff
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          meowth
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          persian
        </span>
      </div>
      <div
        class="sc-bdfBwQ hkRHVq"
      >
        <span>
          farfetchd
        </span>
      </div>
    </div>
    <div
      class="button_wrapper"
    >
      <div
        class="sc-gsTCUz cbbEGE"
      >
        <button
          disabled=""
        >
          previous
        </button>
      </div>
      <div
        class="sc-gsTCUz cbbEGE"
      >
        <button
          disabled=""
        >
          next
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
